### PR TITLE
feat(ngModel): provide ng-empty and ng-not-empty CSS classes

### LINKED ISSUE #10050: Proposal: .ng-prefilled class helper
This would be handy in the following use case:

You want to show validation messages either

* after the form is submitted

* on field touch, **provided the field has already a value**. In this case I'd like the user to notice the validity of the new value right after touch -- instead of waiting until form submission.

So it would work like this (relevant classes mentioned):

1. the field is in a *untouched state*, but has a value: `ng-untouched` `ng-pristine` `ng-prefilled`
2. user *changes* field value: `ng-untouched` `ng-dirty` `ng-prefilled`
3. user *blurs* field: `ng-touched` `ng-dirty` `ng-prefilled`

Does this makes sense to you, guys?

### PR DESCRIPTION
**Pull request by petebacondarwin**
I believe that this feature should be enough to fix #10050

### CODE REVIEW

File: test/ng/directive/ngModelSpec.js [hunk 1]
```diff
@@ -1182,46 +1182,6 @@ describe('ngModel', function() {
...
-        $rootScope.$digest();
-
-        expect(input).toBeValid();
-        expect(input).not.toHaveClass('ng-invalid-maxlength');
-        expect(input).toHaveClass('ng-valid-maxlength');
-        expect(addClassCallCount).toBe(1);
-        expect(removeClassCallCount).toBe(0);
-
-        dealoc(element);
-      }));
```
Comment (gkalpak) [inline test/ng/directive/ngModelSpec.js]: Out of curiosity, why was this removed ?

**Comment by gkalpak** [comment]
I wonder if people will ask for this to be propagated to a parent `FormController`...

**Comment by petebacondarwin** [comment]
I was trying not to mention that :-P

**Comment by Narretz** [comment]
For the original use case, you would then use`ng-pristine` and `ng-not-empty` together?

File: src/ng/directive/ngModel.js [hunk 2]
```diff
@@ -316,6 +318,17 @@ var NgModelController = ['$scope', '$exceptionHandler', '$attrs', '$element', '$
     return isUndefined(value) || value === '' || value === null || value !== value;
   };
 
+  this.$$updateEmptyClasses = function(value) {
+    if (this.$isEmpty(value)) {
```
Comment (Narretz) [inline src/ng/directive/ngModel.js]: Should `this` be `ctrl`?

File: test/ng/directive/ngModelSpec.js [hunk 3]
```diff
@@ -1368,6 +1328,26 @@ describe('ngModel', function() {
   describe('CSS classes', function() {
     var EMAIL_REGEXP = /^[a-z0-9!#$%&'*+\/=?^_`{|}~.-]+@[a-z0-9]([a-z0-9-]*[a-z0-9])?(\.[a-z0-9]([a-z0-9-]*[a-z0-9])?)*$/i;
 
+    it('should set ng-empty or ng-not-empty the view value changes', inject(function($compile, $rootScope, $sniffer) {
```
Comment (Narretz) [inline test/ng/directive/ngModelSpec.js]: .. ng-not-empty (when) the view value changes ..

Comment (Narretz) [inline test/ng/directive/ngModelSpec.js]: Also, line lengths? I never know what the limit should be. 100?

Comment (gkalpak) [inline test/ng/directive/ngModelSpec.js]: Theoretically, yes, they should be 100 chars max :smiley:
See: [CONTRIBUTING.md#-coding-rules][1]

[1]: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#-coding-rules

File: src/ng/directive/ngModel.js (hunk 2 above)
Comment (petebacondarwin) [inline src/ng/directive/ngModel.js]: It doesn't really matter because we have control over when it is called but I agree it is probably better for consistency.

File: test/ng/directive/ngModelSpec.js (hunk 3 above)
Comment (petebacondarwin) [inline test/ng/directive/ngModelSpec.js]: Our jshint rules are now set to 200 for line length :-)

**Comment by petebacondarwin** [comment]
Thanks for the reviews @Narretz and @gkalpak. I added a commit.

File: test/ng/directive/ngModelSpec.js (hunk 3 above)
Comment (gkalpak) [inline test/ng/directive/ngModelSpec.js]: Pure evil :-P

**Comment by petebacondarwin** [comment]
@Narretz  and @gkalpak can you take another look?

File: src/ng/directive/ngModel.js [hunk 4]
```diff
@@ -316,6 +318,17 @@ var NgModelController = ['$scope', '$exceptionHandler', '$attrs', '$element', '$
     return isUndefined(value) || value === '' || value === null || value !== value;
   };
 
+  this.$$updateEmptyClasses = function(value) {
```
Comment (Narretz) [inline src/ng/directive/ngModel.js]: This could be simplified by always using ctrl.$viewValue instead of passing it as an arg. It should only be the viewValue anyway. You'll have to change the order on line https://github.com/angular/angular.js/pull/12848/files#diff-ae1963e24d6328c3aa019c0c1b9f4432R851 though, because otherwise it'll use the old viewValue

**Comment by Narretz** [comment]
Regarding my question
> For the original use case, you would then use ng-pristine and ng-not-empty together?

because only ng-empty doesn't work if I understood the issue correctly.

**Comment by petebacondarwin** [comment]
@Narretz: Yes, you would use a combination of the CSS classes for the original use case, but this seemed reasonable and intuitive to me.

File: src/ng/directive/ngModel.js (hunk 4 above)
Comment (petebacondarwin) [inline src/ng/directive/ngModel.js]: Actually it would have to use `$$lastCommittedViewValue` and we would have to change the order as you suggest. I am not that keen on doing this though as I don't feel that not passing in an argument makes it that much simpler.

**Comment by petebacondarwin** [comment]
Thanks for the review @Narretz.
So assuming that we don't simplify the method and that the original use case is covered by a combination of CSS classes. Is this PR ready to merge?

**Comment by Narretz** [comment]
Yep, feel free to merge!
Am 21.09.2015 22:45 schrieb "Pete Bacon Darwin" <notifications@github.com>:
